README.md
# RestClient for Arduino

HTTP Request library for Arduino.

This library is derived almost entirely from the great work done here: https://github.com/csquared/arduino-restclient

## Examples

The library comes with a number of example sketches. See File > Examples > PlugPlayREST
within the Arduino application.

## Limitations

## Compatible Hardware

The library uses the Arduino Ethernet Client api for interacting with the
underlying network hardware. This means it Just Works with a growing number of
boards and shields, including:

 - Arduino Ethernet
 - Arduino Ethernet Shield
 - Arduino YUN – use the included `YunClient` in place of `EthernetClient`, and
   be sure to do a `Bridge.begin()` first
 - Arduino WiFi Shield
 - Sparkfun WiFly Shield – [library](https://github.com/dpslwk/WiFly)
 - TI CC3000 WiFi - [library](https://github.com/sparkfun/SFE_CC3000_Library)
 - Intel Galileo/Edison
 - ESP8266

## Install

Clone (or download and unzip) the repository to `~/Documents/Arduino/libraries`
where `~/Documents/Arduino` is your sketchbook directory.

    > cd ~/Documents/Arduino
    > mkdir libraries
    > cd libraries
    > git clone https://github.com/plugplayco/plugplay-arduino-rest.git PlugPlayREST

## Usage

### Include

You need to have the `Ethernet` library already included.

```c++
#include <Ethernet.h>
#include <SPI.h>
#include "PlugPlayREST.h"
```

### PlugPlayREST(host/ip, [port])

Constructor to create an PlugPlayREST object to make requests against.

Use domain name and default to port 80:
```c++
PlugPlayREST client("plugplay.co");
```

Use a local IP and an explicit port:
```c++
PlugPlayREST client("192.168.1.50",5000);
```
### setAuth( const char* userKey, const char* boardId)

Set authentication to connect to PlugPlay

Arguments: 
- *userKey* - user Key of your user account.
- *boardId* - board Id of board you want to GET/PUT to.

### dhcp()

Sets up `EthernetClient` with a mac address of `DEADBEEFFEED`. Returns `true` or `false` to indicate if setting up DHCP
was successful or not

```c++
  client.dhcp()
```

Note: you can have multiple PlugPlayREST objects but only need to call
this once.

Note: if you have multiple Arduinos on the same network, you'll need
to give each one a different mac address.

### begin(byte mac[])

It just wraps the `EthernetClient` call to `begin` and DHCPs.
Use this if you need to explicitly set the mac address.

```c++
  byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };
  if (client.begin(mac) == 0) {
     Serial.println("Failed to configure Ethernet using DHCP");
  }
```

### Manual Ethernet Setup

You can skip the above methods and just configure the EthernetClient yourself:

```c++
  byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };
  //the IP address for the shield:
  byte ip[] = { 192, 168, 2, 11 };
  Ethernet.begin(mac,ip);
```

```c++
  byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };
  Ethernet.begin(mac);
```

This is especially useful for debugging network connection issues.

### char* createMsg(String dvName, float data0)
### char* createMsg(String dvName, float data0, float data1)
### char* createMsg(String dvName, float data0, float data1, float data2)

Create a message to send a PUT request

Arguments: 
- *dvName* - device name
- *data0* - first data
- *data1* - second data
- *data2* - third data

## RESTful methods

All methods return an HTTP status code or 0 if there was an error.

### int getPlugPlay(const char* topic)
### int getPlugPlay(const char* topic, String* response)

Start making requests!

```c++
int statusCode = client.getPlugPlay("inTopic");
```

Pass in a string *by reference* for the response:
```c++
String response = "";
int statusCode = client.getPlugPlay("inTopic", &response);
```

### int putPlugPlay(const char* topic, const char* body)
### int putPlugPlay(const char* topic, const char* body, String* response)

```c++
int statusCode = 0;
char* msg = client.createMsg("arduino", 0, 1, 2);
statusCode = client.putPlugPlay("outTopic", msg);
String response = "";
statusCode = client.putPlugPlay("outTopic", msg, &response);
```

### String getName(const char* payload)

Get device name from a received response

```c++
String response = "";
int statusCode = client.getPlugPlay("inTopic", &response);
client.getName(response.c_str());
```

### float getData(const char* payload, const char* type)

Get data from a received response

```c++
String response = "";
int statusCode = client.getPlugPlay("inTopic", &response);
// Get the first data
client.getData(response.c_str(),"data0");
// Get the second data
client.getData(response.c_str(),"data1");
// Get the third data
client.getData(response.c_str(),"data2");
```

### Other functions

You can refer here: https://github.com/csquared/arduino-restclient

## Debug Mode

If you're having trouble, you can always open `RestClient.cpp` and throw at the top:

```c++
#define HTTP_DEBUG
```

Everything happening in the client will get printed to the Serial port.

## License

This code is released under the MIT License.


CHANGES.txt
/************************ PlugPlayREST ***********************/

1.0
    * Add ArduinoJson Lib at commit bff77ab 
    * Add PlugPlayREST class
    * Edit examples code
    * Edit info files

/************************ PubSubClient ***********************/

2.4
   * Add MQTT_SOCKET_TIMEOUT to prevent it blocking indefinitely
     whilst waiting for inbound data
   * Fixed return code when publishing >256 bytes
   
2.3
   * Add publish(topic,payload,retained) function

2.2
   * Change code layout to match Arduino Library reqs

2.1
   * Add MAX_TRANSFER_SIZE def to chunk messages if needed
   * Reject topic/payloads that exceed MQTT_MAX_PACKET_SIZE

2.0
   * Add (and default to) MQTT 3.1.1 support
   * Fix PROGMEM handling for Intel Galileo/ESP8266
   * Add overloaded constructors for convenience
   * Add chainable setters for server/callback/client/stream
   * Add state function to return connack return code

1.9
   * Do not split MQTT packets over multiple calls to _client->write()
   * API change: All constructors now require an instance of Client
      to be passed in.
   * Fixed example to match 1.8 api changes - dpslwk
   * Added username/password support - WilHall
   * Added publish_P - publishes messages from PROGMEM - jobytaffey

1.8
    * KeepAlive interval is configurable in PubSubClient.h
    * Maximum packet size is configurable in PubSubClient.h
    * API change: Return boolean rather than int from various functions
    * API change: Length parameter in message callback changed
       from int to unsigned int
    * Various internal tidy-ups around types
1.7
    * Improved keepalive handling
    * Updated to the Arduino-1.0 API
1.6
    * Added the ability to publish a retained message

1.5
    * Added default constructor
    * Fixed compile error when used with arduino-0021 or later

1.4
    * Fixed connection lost handling

1.3
    * Fixed packet reading bug in PubSubClient.readPacket

1.2
    * Fixed compile error when used with arduino-0016 or later


1.1
    * Reduced size of library
    * Added support for Will messages
    * Clarified licensing - see LICENSE.txt


1.0
    * Only Quality of Service (QOS) 0 messaging is supported
    * The maximum message size, including header, is 128 bytes
    * The keepalive interval is set to 30 seconds
    * No support for Will messages


src/RestClient.h
#include <Arduino.h>
#include <SPI.h>
#include <Ethernet.h>

class RestClient {

  public:
    RestClient(const char* host);
    RestClient(const char* _host, int _port);

    //Client Setup
    bool dhcp();
    int begin(byte*);

    //Generic HTTP Request
    int request(const char* method, const char* path,
                const char* body, String* response);
    // Set a Request Header
    void setHeader(const char*);
    // Set Content-Type Header
    void setContentType(const char*);

    // GET path
    int get(const char*);
    // GET path and response
    int get(const char*, String*);

    // POST path and body
    int post(const char* path, const char* body);
    // POST path and body and response
    int post(const char* path, const char* body, String*);

    // PUT path and body
    int put(const char* path, const char* body);
    // PUT path and body and response
    int put(const char* path, const char* body, String*);

    // DELETE path
    int del(const char*);
    // DELETE path and body
    int del(const char*, const char*);
    // DELETE path and response
    int del(const char*, String*);
    // DELETE path and body and response
    int del(const char*, const char*, String*);

  protected:
    EthernetClient client;
    int readResponse(String*);
    void write(const char*);
    const char* host;
    int port;
    int num_headers;
    const char* headers[10];
	  const char* contentType;
};


src/PlugPlayREST.h
#ifndef PlugPlayREST_h
#define PlugPlayREST_h

#include <RestClient.h>
#include <ArduinoJson.h>

// MQTT_MAX_PAYLOAD_SIZE : Maximum payload size inside a message
#ifndef MQTT_MAX_PAYLOAD_SIZE
#define MQTT_MAX_PAYLOAD_SIZE 100
#endif

class PlugPlayREST: public RestClient {
    private:
        const char* _userKey; 
        const char* _boardId;
        String _authGetTopic, _authPutTopic;
        char payload[MQTT_MAX_PAYLOAD_SIZE];
    public:
        using RestClient::RestClient;
        PlugPlayREST& setAuth( const char* userKey, const char* boardId);
        int getPlugPlay(const char* topic);
    	int getPlugPlay(const char* topic, String* response);
        int putPlugPlay(const char* topic, const char* body);
		int putPlugPlay(const char* topic, const char* body, String* response);
        
        char* createMsg(String dvName, float data0);
        char* createMsg(String dvName, float data0, float data1);
        char* createMsg(String dvName, float data0, float data1, float data2);
        
        String   getName(const char* payload);
        float    getData(const char* payload, const char* type);
};

#endif

src/PlugPlayREST.cpp
#include "PlugPlayREST.h"

PlugPlayREST& PlugPlayREST::setAuth( const char* userKey, const char* boardId){
    this->_userKey = userKey;
    this->_boardId = boardId;
    return *this;
}

int PlugPlayREST::getPlugPlay(const char* topic){
	_authGetTopic = "/api/rest/" + String(_userKey) + "&" + String(_boardId) + "&" + String(topic);
	return request("GET", _authGetTopic.c_str(), NULL, NULL);
}

int PlugPlayREST::getPlugPlay(const char* topic, String* response){
	_authGetTopic = "/api/rest/" + String(_userKey) + "&" + String(_boardId) + "&" + String(topic);
	return request("GET", _authGetTopic.c_str(), NULL, response);
}

int PlugPlayREST::putPlugPlay(const char* topic, const char* body){
	_authPutTopic = "/api/rest/" + String(_userKey) + "&" + String(_boardId) + "&" + String(topic);
	return request("PUT", _authPutTopic.c_str(), body, NULL);
}

int PlugPlayREST::putPlugPlay(const char* topic, const char* body, String* response){
	_authPutTopic = "/api/rest/" + String(_userKey) + "&" + String(_boardId) + "&" + String(topic);
	return request("PUT", _authPutTopic.c_str(), body, response);
}

char* PlugPlayREST::createMsg(String dvName, float data0){    
    return createMsg(dvName, data0, 0, 0);
}

char* PlugPlayREST::createMsg(String dvName, float data0, float data1){    
    return createMsg(dvName, data0, data1, 0);
}

char* PlugPlayREST::createMsg(String dvName, float data0, float data1, float data2){    
    StaticJsonBuffer<MQTT_MAX_PAYLOAD_SIZE> jsonBuffer;
    JsonObject& root = jsonBuffer.createObject(); 
    root["device"] = dvName;
    root["data0"] = data0;
    root["data1"] = data1;
    root["data2"] = data2;
    root.printTo(payload, MQTT_MAX_PAYLOAD_SIZE);
    return payload;
}

String PlugPlayREST::getName(const char* payload){
    StaticJsonBuffer<MQTT_MAX_PAYLOAD_SIZE> jsonBuffer;
    JsonObject& root = jsonBuffer.parseObject(payload);
    if (!root.success()) {
        Serial.println("parseMsg() failed");
        return "";
    }
    return root["device"];
} 

float PlugPlayREST::getData(const char* payload, const char* type){
    StaticJsonBuffer<MQTT_MAX_PAYLOAD_SIZE> jsonBuffer;
    JsonObject& root = jsonBuffer.parseObject(payload);
    if (!root.success()) {
        Serial.println("parseMsg() failed");
        return 0;
    }    
    return root[type];
}

src/RestClient.cpp
#include "RestClient.h"

#ifdef HTTP_DEBUG
#define HTTP_DEBUG_PRINT(string) (Serial.print(string))
#endif

#ifndef HTTP_DEBUG
#define HTTP_DEBUG_PRINT(string)
#endif

RestClient::RestClient(const char* _host){
  host = _host;
  port = 80;
  num_headers = 0;
  contentType = "application/x-www-form-urlencoded";	// default
}

RestClient::RestClient(const char* _host, int _port){
  host = _host;
  port = _port;
  num_headers = 0;
  contentType = "application/x-www-form-urlencoded";	// default
}

bool RestClient::dhcp(){
  byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };
  if (begin(mac) == 0) {
    Serial.println("Failed to configure Ethernet using DHCP");
    return false;
  }

  //give it time to initialize
  delay(1000);
  return true;
}

int RestClient::begin(byte mac[]){
  return Ethernet.begin(mac);
  //give it time to initialize
  delay(1000);
}

// GET path
int RestClient::get(const char* path){
  return request("GET", path, NULL, NULL); 
}

//GET path with response
int RestClient::get(const char* path, String* response){
  return request("GET", path, NULL, response);
}

// POST path and body
int RestClient::post(const char* path, const char* body){
  return request("POST", path, body, NULL);
}

// POST path and body with response
int RestClient::post(const char* path, const char* body, String* response){
  return request("POST", path, body, response);
}

// PUT path and body
int RestClient::put(const char* path, const char* body){
  return request("PUT", path, body, NULL);
}

// PUT path and body with response
int RestClient::put(const char* path, const char* body, String* response){
  return request("PUT", path, body, response);
}

// DELETE path
int RestClient::del(const char* path){
  return request("DELETE", path, NULL, NULL);
}

// DELETE path and response
int RestClient::del(const char* path, String* response){
  return request("DELETE", path, NULL, response);
}

// DELETE path and body
int RestClient::del(const char* path, const char* body ){
  return request("DELETE", path, body, NULL);
}

// DELETE path and body with response
int RestClient::del(const char* path, const char* body, String* response){
  return request("DELETE", path, body, response);
}

void RestClient::write(const char* string){
  HTTP_DEBUG_PRINT(string);
  client.print(string);
}

void RestClient::setHeader(const char* header){
  headers[num_headers] = header;
  num_headers++;
}

void RestClient::setContentType(const char* contentTypeValue){
  contentType = contentTypeValue;
}

// The mother- generic request method.
//
int RestClient::request(const char* method, const char* path,
                  const char* body, String* response){

  HTTP_DEBUG_PRINT("HTTP: connect\n");

  if(client.connect(host, port)){
    HTTP_DEBUG_PRINT("HTTP: connected\n");
    HTTP_DEBUG_PRINT("REQUEST: \n");
    // Make a HTTP request line:
    write(method);
    write(" ");
    write(path);
    write(" HTTP/1.1\r\n");
    for(int i=0; i<num_headers; i++){
      write(headers[i]);
      write("\r\n");
    }
    write("Host: ");
    write(host);
    write("\r\n");
    write("Connection: close\r\n");

    if(body != NULL){
      char contentLength[30];
      sprintf(contentLength, "Content-Length: %d\r\n", strlen(body));
      write(contentLength);

	  write("Content-Type: ");
	  write(contentType);
	  write("\r\n");
    }

    write("\r\n");

    if(body != NULL){
      write(body);
      write("\r\n");
      write("\r\n");
    }

    //make sure you write all those bytes.
    delay(100);

    HTTP_DEBUG_PRINT("HTTP: call readResponse\n");
    int statusCode = readResponse(response);
    HTTP_DEBUG_PRINT("HTTP: return readResponse\n");

    //cleanup
    HTTP_DEBUG_PRINT("HTTP: stop client\n");
    num_headers = 0;
    client.stop();
    delay(50);
    HTTP_DEBUG_PRINT("HTTP: client stopped\n");

    return statusCode;
  }else{
    HTTP_DEBUG_PRINT("HTTP Connection failed\n");
    return 0;
  }
}

int RestClient::readResponse(String* response) {

  // an http request ends with a blank line
  boolean currentLineIsBlank = true;
  boolean httpBody = false;
  boolean inStatus = false;

  char statusCode[4];
  int i = 0;
  int code = 0;

  if(response == NULL){
    HTTP_DEBUG_PRINT("HTTP: NULL RESPONSE POINTER: \n");
  }else{
    HTTP_DEBUG_PRINT("HTTP: NON-NULL RESPONSE POINTER: \n");
  }

  HTTP_DEBUG_PRINT("HTTP: RESPONSE: \n");
  while (client.connected()) {
    HTTP_DEBUG_PRINT(".");

    if (client.available()) {
      HTTP_DEBUG_PRINT(",");

      char c = client.read();
      HTTP_DEBUG_PRINT(c);

      if(c == ' ' && !inStatus){
        inStatus = true;
      }

      if(inStatus && i < 3 && c != ' '){
        statusCode[i] = c;
        i++;
      }
      if(i == 3){
        statusCode[i] = '\0';
        code = atoi(statusCode);
      }

      if(httpBody){
        //only write response if its not null
        if(response != NULL) response->concat(c);
      }
      else
      {
          if (c == '\n' && currentLineIsBlank) {
            httpBody = true;
          }

          if (c == '\n') {
            // you're starting a new line
            currentLineIsBlank = true;
          }
          else if (c != '\r') {
            // you've gotten a character on the current line
            currentLineIsBlank = false;
          }
      }
    }
  }

  HTTP_DEBUG_PRINT("HTTP: return readResponse3\n");
  return code;
}
